library/src/trips/index.ts

```diff
@@ -0,0 +1,170 @@
+      // we could add the possibility to specify the country of your
+      // city (with a syntax like 'new york usa paris france')
+      // so it seems more future-proof to take the population into
+      // account only here.
+      const score =
+        distance(formattedName, city.formattedName) -
+        city.population / 50_000_000;
+      if (score < bestCity.score) {
+        return {
+          score,
+          city,
+        };
+      }
+
+      return bestCity;
+    },
+    { city: filteredCities[0], score: Infinity }
+  );
+};
+
+const findCities = (
+  query: string
+): { cities: [City, City]; score: number } | undefined => {
+  // Our query is something like "san fransisco new york"
+  // There can be only two cities, one at the start, and
+  // one at the end.
+  // All the magic is to split the string at the right place.
+  // Which space separate the two cities.
+  // This is what this function is doing.
+
+  const words = query.split(" ");
+  const wordsCount = words.length;
+
+  if (wordsCount < 2) {
+    return;
+  }
+
+  let bestScore = Infinity;
+  let bestCities;
+
+  for (let i = 1; i < wordsCount; i += 1) {
+    const firstWord = words.slice(0, i).join(" ");
+    const secondWord = words.slice(i, wordsCount).join(" ");
+
+    const firstCity = findBestCity(firstWord);
+    const secondCity = findBestCity(secondWord);
+
+    const score = firstCity.score + secondCity.score;
+    if (score < bestScore) {
+      bestScore = score;
+      bestCities = [firstCity.city, secondCity.city];
+    }
+  }
+
+  return { cities: bestCities as any as [City, City], score: bestScore };
+};
+
+const deg2rad = (deg: number) => {
+  return deg * (Math.PI / 180);
+};
+
+const getDistanceFromLatLonInKm = (
+  [lon1, lat1]: LngLat,
+  [lon2, lat2]: LngLat
+) => {
+  // https://stackoverflow.com/questions/18883601/function-to-calculate-distance-between-two-coordinates
+  const R = 6371; // Radius of the earth in km
+  const dLat = deg2rad(lat2 - lat1); // deg2rad below
+  const dLon = deg2rad(lon2 - lon1);
+  const a =
+    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
+    Math.cos(deg2rad(lat1)) *
+      Math.cos(deg2rad(lat2)) *
+      Math.sin(dLon / 2) *
+      Math.sin(dLon / 2);
+  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+  const d = R * c; // Distance in km
+  return d;
+};
+
+const CO2EQ_FLIGHT_SOURCES = ["https://www.carbonindependent.org/22.html"];
+const CO2EQ_KG_PER_KM_FLIGHT = 0.158;
+
+export const getTripFromQuery = (
+  query: string
+): { trip: Trip; score: number } | undefined => {
+  const citiesResult = findCities(query);
+  if (!citiesResult) {
+    return;
+  }
+
+  const {
+    cities: [origin, destination],
+    score,
+  } = citiesResult;
+
+  const distanceInKm = getDistanceFromLatLonInKm(
+    origin.loc.coordinates,
+    destination.loc.coordinates
+  );
+
+  const carbonInKg = distanceInKm * CO2EQ_KG_PER_KM_FLIGHT;
+
+  const trip = Trip.build({
+    kind: "trip",
```

Comment:
This can be done in the `build()` method

Reply:
Ok, I'll try to do that

Reply:
I thought I succeeded but no, I tried different things but either `kind` disappears when serializing the object to JSON, either I have TS issues.

If you have the solution, happy to hear it!